pkg/controllers/cronfederatedhpa/cronfederatedhpa_controller.go

```diff
@@ -0,0 +1,206 @@
+			return controllerruntime.Result{}, nil
+		}
+
+		klog.Error("Failed to get CronFederatedHPA(%s):", req.NamespacedName.String(), err)
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			c.EventRecorder.Event(cronFHPA, corev1.EventTypeWarning, "ProcessCronFederatedHPAFailed", err.Error())
+		}
+	}()
+
+	// If this CronFHPA is deleted, remove all the cron executors
+	if !cronFHPA.DeletionTimestamp.IsZero() {
+		c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+		return controllerruntime.Result{}, nil
+	}
+
+	origRuleSets := sets.New[string]()
+	for _, history := range cronFHPA.Status.ExecutionHistories {
+		origRuleSets.Insert(history.RuleName)
+	}
+
+	// If scale target is updated, stop all the rule executors, and next steps will create the new executors
+	if c.CronHandler.CronFHPAScaleTargetRefUpdates(req.NamespacedName.String(), cronFHPA.Spec.ScaleTargetRef) {
+		c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+	}
+
+	c.CronHandler.AddCronExecutorIfNotExist(req.NamespacedName.String())
+
+	newRuleSets := sets.New[string]()
+	for _, rule := range cronFHPA.Spec.Rules {
+		if err = c.processCronRule(cronFHPA, rule); err != nil {
+			return controllerruntime.Result{}, err
+		}
+		newRuleSets.Insert(rule.Name)
+	}
+
+	// If rule is deleted, remove the rule executor from the handler
+	for name := range origRuleSets {
+		if newRuleSets.Has(name) {
+			continue
+		}
+		c.CronHandler.StopRuleExecutor(req.NamespacedName.String(), name)
+		if err = c.removeCronFHPAHistory(cronFHPA, name); err != nil {
+			return controllerruntime.Result{}, err
+		}
+	}
+
+	return controllerruntime.Result{}, nil
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *CronFHPAController) SetupWithManager(mgr controllerruntime.Manager) error {
+	c.CronHandler = NewCronHandler(mgr.GetClient(), mgr.GetEventRecorderFor(ControllerName))
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&autoscalingv1alpha1.CronFederatedHPA{}).
+		WithOptions(controller.Options{RateLimiter: ratelimiterflag.DefaultControllerRateLimiter(c.RateLimiterOptions)}).
+		Complete(c)
+}
+
+// processCronRule processes the cron rule
+func (c *CronFHPAController) processCronRule(cronFHPA *autoscalingv1alpha1.CronFederatedHPA, rule autoscalingv1alpha1.CronFederatedHPARule) error {
+	cronFHPAKey := helper.GetCronFederatedHPAKey(cronFHPA)
+	if ruleOld, exists := c.CronHandler.RuleCronExecutorExists(cronFHPAKey, rule.Name); exists {
+		if equality.Semantic.DeepEqual(ruleOld, rule) {
+			return nil
+		}
+		c.CronHandler.StopRuleExecutor(cronFHPAKey, rule.Name)
+	}
+
+	if !helper.IsCronFederatedHPARuleSuspend(rule) {
+		if err := c.CronHandler.CreateCronJobForExecutor(cronFHPA, rule); err != nil {
+			klog.Error("Fail to start cron for CronFederatedHPA(%s):%v", cronFHPAKey, err)
+			return err
+		}
+	}
+
+	return c.updateRuleNextExecutionTime(cronFHPA, rule)
```

Comment:
I think that when `Suspend` is set to `true`, there is no need to calculate `NextExecutionTime`.


```go 

	if helper.IsCronFederatedHPARuleSuspend(rule) {
            return nil
	}
	if err := c.CronHandler.CreateCronJobForExecutor(cronFHPA, rule); err != nil {
		klog.Error("Fail to start cron for CronFederatedHPA(%s):%v", cronFHPAKey, err)
		return err
	}
	return c.updateRuleNextExecutionTime(cronFHPA, rule)


```

Reply:
updateRuleNextExecutionTime will not add `NextExecutionTime`:
```go
func (c *CronFHPAController) updateRuleNextExecutionTime(cronFHPA *autoscalingv1alpha1.CronFederatedHPA, rule autoscalingv1alpha1.CronFederatedHPARule) error {
	var nextExecutionTime *metav1.Time
	if !helper.IsCronFederatedHPARuleSuspend(rule) {
		// If rule is not suspended, we should set the nextExecutionTime filed, or the nextExecutionTime will be nil
		next, err := helper.GetCronFederatedHPARuleNextExecutionTime(rule)
		if err != nil {
			klog.Errorf("Failed to get next execution time for CronFederatedHPA(%s/%s):%v", cronFHPA.Namespace, cronFHPA.Name, err)
			return err
		}
		nextExecutionTime = &metav1.Time{Time: next}
	}
```

I rename the function to updateRuleHistory